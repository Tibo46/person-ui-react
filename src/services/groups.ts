import * as t from 'io-ts';
import { get } from '../utils/http';

export const Group = t.strict({
  id: t.string,
  name: t.string,
});
export type Group = t.TypeOf<typeof Group>;

export const GroupResponse = t.array(Group);

export type GroupResponse = t.TypeOf<typeof GroupResponse>;
export async function getTypes(): Promise<Group[]> {
  const res = await get(GroupResponse, {
    url: `${window.ENV_DATA.personApiOrigin}/groups`,
  });

  return res;
}
