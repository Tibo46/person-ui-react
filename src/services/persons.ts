import * as t from 'io-ts';
import { get } from '../utils/http';
import { Group } from './groups';

export const Person = t.strict({
  id: t.string,
  name: t.string,
  group: Group,
});
export type Person = t.TypeOf<typeof Person>;

export const SearchResponse = t.array(Person);

export type SearchResponse = t.TypeOf<typeof SearchResponse>;

interface SearchPersonQuery {
  search: string;
}
export async function searchPersons({ query }: { query: SearchPersonQuery }): Promise<Person[]> {
  const searchTerms = query.search ? `?searchTerms=${query.search}` : '';

  const res = await get(SearchResponse, {
    url: `${window.ENV_DATA.personApiOrigin}/persons/search${searchTerms}`,
  });

  return res;
}

interface GetPersonQuery {
  id: string;
}
export async function getPerson({ query }: { query: GetPersonQuery }): Promise<Person> {
  const res = await get(Person, {
    url: `${window.ENV_DATA.personApiOrigin}/persons/${query.id}`,
  });

  return res;
}
