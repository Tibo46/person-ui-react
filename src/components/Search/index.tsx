import React from 'react';
import { useAsync } from 'react-async';
import Typography from '@material-ui/core/Typography';
import { searchPersons, Person } from '../../services/persons';
import Loader from '../../components/Loader';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogActions from '@material-ui/core/DialogActions';
import TextField from '@material-ui/core/TextField';
import SearchIcon from '@material-ui/icons/Search';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Link from '../Link';
import ListItemText from '@material-ui/core/ListItemText';

const useStyles = makeStyles({
  loader: {
    position: 'absolute',
    right: '10px',
    top: '10px',
    width: '30px',
    height: '30px',
  },
  dialogContent: {
    minHeight: '105px',
    paddingTop: '1rem',
  },
  employee: {
    display: 'flex',
    width: '100%',
  },
  employeeName: {
    marginRight: '.5rem',
  },
  employeeGroup: {
    marginLeft: '.5rem',
  },
});

const Search = () => {
  const classes = useStyles();
  const [filter, setFilter] = React.useState('');
  const [open, setOpen] = React.useState(false);

  // @ts-ignore
  const personsResult = useAsync({
    promiseFn: open && filter ? searchPersons : false,
    query: {
      search: filter,
    },
    watch: filter,
  });

  const handleClose = () => {
    setFilter('');
    personsResult.setData([]);
    setOpen(false);
  };

  return (
    <>
      <IconButton onClick={(event) => setOpen(true)}>
        <SearchIcon />
      </IconButton>
      <Dialog
        fullWidth={true}
        open={open}
        onClose={handleClose}
        aria-labelledby="search-dialog-title"
      >
        {personsResult.isLoading && (
          <div className={classes.loader}>
            <Loader />
          </div>
        )}
        <DialogTitle id="search-dialog-title">Search</DialogTitle>
        <DialogContent className={classes.dialogContent}>
          <TextField
            autoComplete="off"
            fullWidth={true}
            variant="outlined"
            label="Search by employee or group name"
            value={filter}
            name="filter"
            onChange={(e) => {
              setFilter(e.target.value);
            }}
          />
          <List>
            {filter &&
              (personsResult.data && personsResult.data.length > 0 ? (
                personsResult.data.map((person: Person, index: number) => (
                  <ListItem button key={`person-${person.id}`}>
                    <Link
                      href={`/employees/details/${person.id}`}
                      className={classes.employee}
                      onClick={handleClose}
                    >
                      <ListItemText primary={person.name} className={classes.employeeName} />
                      <ListItemText
                        primary={person.group?.name}
                        className={classes.employeeGroup}
                      />
                    </Link>
                  </ListItem>
                ))
              ) : (
                <Typography className={`no-data`} id="no-data">
                  No employee found
                </Typography>
              ))}
          </List>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} id="dialog-search-close" color="primary">
            Close
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default Search;
