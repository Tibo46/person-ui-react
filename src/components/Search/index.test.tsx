import React from 'react';
import { act } from 'react-dom/test-utils';
import fetchMock from 'fetch-mock';
import { MemoryRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import { SearchResponse } from '../../services/persons';
import Search from '.';

window.ENV_DATA = {
  personApiOrigin: 'https://mock-api.com',
} as any;

describe('Search component', () => {
  let allPersonsResult: SearchResponse;

  beforeEach(() => {
    allPersonsResult = [
      {
        id: 'employee-1',
        name: 'Walter White',
        groupId: 1,
        group: {
          id: 1,
          name: 'Chemist',
        },
      },
    ];

    fetchMock.get(
      `${window.ENV_DATA.personApiOrigin}/persons/search?searchTerms=Walter`,
      allPersonsResult
    );
    fetchMock.get(
      `${window.ENV_DATA.personApiOrigin}/persons/search?searchTerms=This%20one%20does%20not%20exists`,
      []
    );
  });

  afterEach(() => {
    fetchMock.reset();
  });

  it('renders the search button', async () => {
    const component = renderer.create(<Search />);

    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('opens the modal once search button clicked', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(<Search />);
    });
    component!.update();

    expect(component!.find('mui-dialog').length).toBe(0);

    act(() => {
      component.find('mui-icon-button').at(0).simulate('click');
    });
    component!.update();

    expect(component!.find('mui-dialog').length).toBe(1);
  });

  it('closes the modal on close button click', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(<Search />);
      component.find('mui-icon-button').at(0).simulate('click');
    });
    component!.update();

    expect(component!.find('mui-dialog').length).toBe(1);

    act(() => {
      component.find('#dialog-search-close').at(0).simulate('click');
    });
    component!.update();

    expect(component!.find('mui-dialog').length).toBe(0);
  });

  it('triggers the search when the search input value changes', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(
        <Router>
          <Search />
        </Router>
      );
      component.find('mui-icon-button').at(0).simulate('click');
    });
    component!.update();

    await act(async () => {
      component.find('input[name="filter"]').simulate('change', { target: { value: 'Walter' } });
    });
    component!.update();

    expect(fetchMock.calls()).toHaveLength(1);
    expect(fetchMock.calls()[0][0]).toContain('Walter');
  });

  it('displays no data if no employee match the search terms', async () => {
    let component: ReactWrapper;
    await act(async () => {
      component = mount(<Search />);
      component.find('mui-icon-button').at(0).simulate('click');
    });
    component!.update();

    await act(async () => {
      component
        .find('input[name="filter"]')
        .simulate('change', { target: { value: 'This one does not exists' } });
    });
    component!.update();

    expect(fetchMock.calls()).toHaveLength(1);
    await act(async () => {
      expect(component.find('p#no-data')).toHaveLength(1);
    });
  });
});
