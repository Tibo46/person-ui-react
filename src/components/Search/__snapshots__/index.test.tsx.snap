// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders the search button 1`] = `
<mui-icon-button
  onClick={[Function]}
>
  <mui-icon-search />
</mui-icon-button>
`;
